src/ripple/rpc/handlers/Catalogue.cpp

```diff
@@ -0,0 +1,1139 @@
+        infile.open(filepath.c_str(), std::ios::in | std::ios::binary);
+        if (infile.fail())
+            return rpcError(
+                rpcINTERNAL,
+                "cannot reopen file after hash verification: " +
+                    std::string(strerror(errno)));
+
+        // Skip the header
+        infile.seekg(sizeof(CATLHeader), std::ios::beg);
+    }
+
+    // Set up decompression if needed
+    auto decompStream = std::make_unique<boost::iostreams::filtering_istream>();
+    if (compressionLevel > 0)
+    {
+        JLOG(context.j.info())
+            << "Setting up decompression with level " << (int)compressionLevel;
+        boost::iostreams::zlib_params params((int)compressionLevel);
+        params.window_bits = 15;
+        params.noheader = false;
+        decompStream->push(boost::iostreams::zlib_decompressor(params));
+    }
+    else
+    {
+        JLOG(context.j.info())
+            << "No decompression needed (level 0), using direct input";
+    }
+    decompStream->push(boost::ref(infile));
+
+    uint32_t ledgersLoaded = 0;
+    std::shared_ptr<Ledger> prevLedger;
+    uint32_t expected_seq = header.min_ledger;
+
+    // Process each ledger sequentially
+    while (!decompStream->eof() && expected_seq <= header.max_ledger)
+    {
+        if (context.app.isStopping())
+            return {};
+
+        // Update current ledger
+        UPDATE_CATALOGUE_STATUS(ledgerUpto, expected_seq);
+
+        LedgerInfo info;
+        uint64_t closeTime = -1;
+        uint64_t parentCloseTime = -1;
+        uint32_t closeTimeResolution = -1;
+        uint64_t drops = -1;
+
+        if (!decompStream->read(
+                reinterpret_cast<char*>(&info.seq), sizeof(info.seq)) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(info.hash.data()), 32) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(info.txHash.data()), 32) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(info.accountHash.data()), 32) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(info.parentHash.data()), 32) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(&drops), sizeof(drops)) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(&info.closeFlags),
+                sizeof(info.closeFlags)) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(&closeTimeResolution),
+                sizeof(closeTimeResolution)) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(&closeTime), sizeof(closeTime)) ||
+            !decompStream->read(
+                reinterpret_cast<char*>(&parentCloseTime),
+                sizeof(parentCloseTime)))
+        {
+            JLOG(context.j.warn()) << "Catalogue load expected but could not "
+                                      "read the next ledger header.";
+            break;
```

Comment:
This should jump to error handling somehow, or set an error flag ?
Otherwise, per comment below, it will errantly setLedgerRangePresent for the full header range